### ISSUE TITLE
Make Record2 extend Record1

### ISSUE THREAD

**Issue by adrianmsmith** [jOOQ/jOOQ#4604]
I have the situation that I am fetching a lot of stuff and then calling some function to process that stuff. That function to process the stuff takes a typesafe `Record8<Integer, ...>`, so if I make a mistake in my select or or assessor code the compiler can tell me, and that's brilliant.

However in some code I'm now writing, I need to fetch some additional information (which I would like to do in one select, i.e. joins etc.). I would like to reuse the above code in a typesafe manner. I would add my new fields to the right-hand-side of the select, giving me e.g. a `Record9<Integer, ..... String>` (if my new field is a string).

If e.g. `Record2<Integer, String>` extended `Record1<Integer>` this would all work nicely including the `Integer value1()` method on the object.

**Comment by lukaseder** [comment]
Thanks for your feature request. Many developers who implement tuples in their APIs are lured into establishing type dependencies like the one you suggested. After all, the method names are already there and seem to match. However, this is a terrible mistake because `Tuple1<T1>` and `Tuple2<T1, T2>` are two fundamentally incompatible types (although, they may have a common base type).

One obvious example where this becomes apparent is when using `IN` predicates:

```java
// Compiles:
TABLE1.COL1.in(select(TABLE2.COL1).from(TABLE2))

// Must not compile (but would, after your suggested change):
TABLE1.COL1.in(select(TABLE2.COL1, TABLE2.COL2).from(TABLE2))
```

The signature of `Field<T>.in(Select<? extends Record1<T>>)` currently guarantees tuple compatibility on both sides of the `IN` predicate. This would no longer be the case if `Record2 extends Record1`.

Also, in simpler terms, you may prefer to map `Record1.value1()` to `Record2.value1()`, or more generally:

```java
Record2<T1, T2> r2 = ...
Record1<T1> r1 = r2;
```

But the following pattern might just be as desireable in some cases:

```java
Record2<T1, T2> r2 = ...
Record1<T2> r1 = r2;
```

Or, more "wildly":

```java
Record5<T1, T2, T3, T4, T5> r5 = ...
Record2<T4, T1> r2 = r5;
```

There are a factorial amount of permutations that might "make sense" in the context of such an assignment. The one you suggested is just an accidental "special case".

**TL;DR**: We can't implement that, I'm sorry :)

**Comment by adrianmsmith** [comment]
Understood, thanks for your very clear and helpful explanation. Thanks!

**Comment by lukaseder** [comment]
On a side-note: We may well add useful methods to all `Record[N]` types to help you *convert* them between one another. I'll think of some examples. E.g. we could "split" a `Record5<T1, T2, T3, T4, T5>` at element 3 to produce `Record2<T1, T2>` and a `Record3<T3, T4, T5>`...

**Comment by adrianmsmith** [comment]
I solved it by using the `aRecord2.into(MY_TABLE.MY_FIELD)` which converted from `Record2` to `Record1` however that would fail at runtime if i change the select to not include MY_FIELD, whereas such a "split" as you propose would probably allow such things to be detected at compile-time. Thanks again!
